src/DataProviders/SqlDataProvider/SqlDataProvider.ts

```diff
@@ -1,61 +1,83 @@
-            .then((pool) => {
-                return pool.request()
-                    .query("select * from Hymns_Seasons");
-            }).then((result) => {
-                assert.notEqual(result, null);
-                assert.notEqual(result.recordsets, null);
-                assert.notEqual(result.recordsets.length, 0);
+    public async getSeasonList(): Promise<SeasonInfo[]> {
+        try {
+            const connection = await this._cp;
+            const result = await connection.request()
+                .query("select * from Hymns_Seasons");
 
-                const seasons: SeasonInfo[] = result.recordsets[0]
-                    .map((row) => SqlDataProvider._convertSeasonDbItemToSeasonInfo(row));
-                return seasons;
-            }).catch((err) => {
-                console.log(err);
-            });
+            assert.ok(result);
+            assert.ok(result.recordsets);
+            assert.ok(result.recordsets.length);
+            assert.notEqual(result.recordsets.length, 0);
+
+            const seasons: SeasonInfo[] = result.recordsets[0]
+                .map((row) => SqlDataProvider._convertSeasonDbItemToSeasonInfo(row));
+            return seasons;
+        } catch (e) {
+            console.log(e);
+            throw e;
+        }
     }
 
-    public getSeason(seasonId: string): Promise<void | SeasonInfo> {
-        const cp = new ConnectionPool(this._config);
-        return cp.connect()
-            .then((pool) => {
-                return pool.request()
-                    .query(`select * from Hymns_Seasons where ID = ` + seasonId);
-            }).then((result) => {
-                assert.notEqual(result, null);
-                assert.notEqual(result.recordsets, null);
-                assert.notEqual(result.recordsets.length, 0);
-                const row = result.recordsets[0][0];
-                return SqlDataProvider._convertSeasonDbItemToSeasonInfo(row);
-            }).catch((err) => {
-                console.log(err);
-            });
+    public async getSeason(seasonId: string): Promise<SeasonInfo> {
+        try {
+            const connection = await this._cp;
+            const result = await connection.request()
+                .query(`select * from Hymns_Seasons where ID = ` + seasonId);
+
+            assert.ok(result);
+            assert.ok(result.recordsets);
+            assert.ok(result.recordsets.length);
+            assert.notEqual(result.recordsets.length, 0);
+            const row = result.recordsets[0][0];
+            return SqlDataProvider._convertSeasonDbItemToSeasonInfo(row);   
+        } catch (e) {       
+            console.log(e);
+            throw e;
+        }
     }
 
-    private static _convertSeasonDbItemToSeasonInfo(seasonDbItem: any): SeasonInfo {
+    private static _convertSeasonDbItemToSeasonInfo(seasonDbItem: HazzatDbSchema.Season): SeasonInfo {
+        if (isNullOrUndefined(seasonDbItem)) {
+            return null;
```

Comment:
do you really want to return null in such a case? wouldn't you rather throw so you can fail fast? it'd cause an exception somewhere else anyway

Reply:
Reworked all error handling